torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -694,7 +1035,61 @@ def _build_param_buckets(self) -> None:
                         offset = offset_next
-                self._buckets[dev_i].append(bucket)
+                self._buckets[dev_i].append(bucket)  # type: ignore[arg-type]
+
+    def _build_ddp_param_buckets(self) -> None:
+        r"""
+        Builds parameter buckets if ``overlap_with_ddp`` so that for each
+        device that stores this rank's parameters, there is a :class:`list` of
+        buckets (represented as tensors) containing the parameters on that
+        device that are assigned to the rank in the parameter update
+        partition and grouped following the :class:`DistributedDataParallel`
+        gradient buckets.
+
+        This method should only be called during the delayed initialization
+        when ``overlap_with_ddp=True``.
+
+        .. warning::
+            The current implementation assumes that all of the parameters in a
+            bucket are of the same dense type when allocating the bucket's
+            tensor.
+
+        .. warning::
+            If the model parameters are stored across more than one device,
+            then the storage partitioning must be the same across all
+            processes in order for parameter synchronization to work.
+        """
+        assert self._overlap_with_ddp, \
+            "`_build_ddp_param_buckets()` should only be called when " \
+            "`overlap_with_ddp=True`"
+
+        num_devices = len(self._device_to_buckets)
+        self._buckets = [[{} for _ in range(self.world_size)] for _ in range(num_devices)]  # type: ignore[assignment]
+
+        for dev_idx, (device, ddp_buckets_per_rank) in enumerate(self._device_to_buckets.items()):
+            self._device_to_device_index[device] = dev_idx
+            for rank, ddp_buckets in enumerate(ddp_buckets_per_rank):
+                for ddp_bucket in ddp_buckets:
+                    bucket_index = ddp_bucket.bucket_index  # type: ignore[attr-defined]
+                    params = ddp_bucket.params              # type: ignore[attr-defined]
+                    bucket_size = 0
+                    dtype = None
+                    for param in params:
+                        assert _is_trainable(param), \
+                            "Model parameter corresponding to a gradient in " \
+                            "a DDP bucket should require a gradient"
+                        bucket_size += param.numel()
+                        dtype = param.dtype  # assumes all same dtype
+                    assert bucket_size > 0
+                    bucket = torch.empty(bucket_size, dtype=dtype, device=device)
+                    offset = 0
+                    # Construct the bucket (assuming all dense and same dtype)
+                    for param in params:
+                        offset_next = offset + param.numel()
+                        bucket[offset:offset_next].copy_(param.data.flatten())
+                        param.data = bucket[offset:offset_next].view_as(param.data)
```

Comment:
One issue with using `set_()` is that we must set `requires_grad=False` and later reset it to `True` since otherwise we get the error saying that `a leaf Variable that requires grad is being used in an in-place operation.` More importantly, I am unable to get it to manage the strides correctly. As far as I can see, I do not think it will be any simpler to use `set_()` for this copying.